performant-software/dm-2
# Add interface for viewing and creating links to popups and sidebar

### What does this PR do?
- Adds the link inspector view, which can be accessed for any document or highlight by clicking on the document's link icon, clicking the highlight, or double-clicking either's summary inside another link inspector.
- Enables the creation of links for existing resources by dragging highlight or document summaries onto the drop zones of another highlight/document's link inspector.
- Enables the creation of documents from inside the link inspector, attaching the new document to the source of link inspector instead of to the table of contents.
- Improves text editor highlight interaction, including highlight deletion with confirmation, text changes reflected in excerpts, and duplicating link data for pasted highlights.

### What issues does it address?
- Closes #19 
- Closes #20 
- Closes #21 
- Closes #22 

### How to test
- In the [review app](https://dm-2-staging-pr-37.herokuapp.com/), open a new or existing project.
- Add some text documents with text content and highlights.
- Clicking on a highlight should bring up a link inspector popup, positioned just below where you clicked. The inspector should show a summary of the highlight you clicked, outlined at the top of the element, and below it a list of documents/highlights linked to that highlight, followed by a drop target to make new links and a button to add new linked documents.
- Clicking on the link icon to the left of a document viewer's title should bring up the same kind of popup, showing the summary and links for the document itself.
- Link inspector popups should be repositionable above the document grid via clicking and dragging their title bars. They should be resizable via the lower right corner. When the title bar is clicked or dragged, that popup should move to the front.
- Double-clicking any highlight/document summary (whether at the top of a link inspector or one of the link items) should switch the project sidebar from showing a table of contents to showing a link inspector for that highlight/document. Like the popups, this view should show a yellow backdrop. You should be able to return to the table of contents view via a button at the top of the sidebar.
- With two or more link inspectors open in popups and/or the sidebar, create a new link by dragging the top summary for one highlight/document onto another's drop target. Both link inspectors (where you dragged from and where you dropped) should refresh to show the new link in their lists.
- Click on the Add New Linked Document button at the bottom of a link inspector and select Text. A new text document viewer should appear in the grid. A link to that document should appear in the link inspector, but the document should not appear in the table of contents. Edit its title to replace the default Untitled Document value - the title should update in the link inspector list after a slight delay. Try this from both a link inspector popup and the link inspector view of the sidebar.
- Experiment with creating text highlights in various ways, e.g. using the toggle highlight button on a selection, toggling with no selection and then typing, pasting in existing highlights from the same or another document. Highlights should always bring a link inspector up when clicked, and should always show the correct excerpt in that or any other inspector referencing them. Try deleting highlights and modifying their text content by partially deleting or typing within them.
- When one or more highlights would fully disappear, a dialog box should appear and ask you to confirm the change. Clicking Cancel should return the text document to its state before whatever change would have removed/replaced the highlight(s).
